deep_qa/training/models.py

```diff
@@ -28,6 +35,62 @@ def _summary_with_mask_info(self):
         flattened_layers = getattr(self, 'flattened_layers', self.layers)
         print_summary_with_masking(flattened_layers, getattr(self, 'container_nodes', None))
 
+    @overrides
+    def compile(self, params: Params):  # pylint: disable=arguments-differ
+        # pylint: disable=attribute-defined-outside-init
+        """
+        The only reason we are overriding this method is because keras automatically wraps
+        our tensorflow optimiser in a keras wrapper, which we don't want. We override the
+        only method in ``Model`` which uses this attribute, ``_make_train_function``, which
+        raises an error if compile is not called first.
+
+        As we move towards using a Tensorflow first optimisation loop, more things will be
+        added here which add functionality to the way Keras runs tensorflow Session calls.
+
+        """
+        optimizer = params.get('optimizer')
+        self.gradient_clipping = params.pop("gradient_clipping", None)
+        print(self.gradient_clipping.as_dict())
+        super(DeepQaModel, self).compile(**params.as_dict())
+        self.optimizer = optimizer
+
+    @overrides
+    def _make_train_function(self):
+        # pylint: disable=attribute-defined-outside-init
+        """
+        We override this method so that we can use tensorflow optimisers directly.
+        This is desirable as tensorflow handles gradients of sparse tensors efficiently.
+        """
+        if not hasattr(self, 'train_function'):
+            raise RuntimeError('You must compile your model before using it.')
+        if self.train_function is None:
+            inputs = self._feed_inputs + self._feed_targets + self._feed_sample_weights
+            if self.uses_learning_phase and not isinstance(K.learning_phase(), int):
+                inputs += [K.learning_phase()]
+
+            # Here we override Keras to use tensorflow optimizers directly.
+            self.global_step = K.variable(0., name='global_step')
+            grads = tensorflow.gradients(self.total_loss, self._collected_trainable_weights)
+            if self.gradient_clipping is not None:
+                # Don't pop from the gradient clipping dict here as
+                # when we load models we need it to still be there.
```

Comment:
You mean if you call `fit()` twice, you need it to still be there.  The issue with this is that the parameter won't get logged, though.  Can you call `self.gradient_clipping.as_dict()`, then read from that?